activerecord/lib/active_record/testing/query_assertions.rb

```diff
@@ -7,28 +7,102 @@ module QueryAssertions
       #
       #   assert_queries(1) { Post.first }
       #
-      # If the +:matcher+ option is provided, only queries that match the matcher are counted.
-      #
-      #   assert_queries(1, matcher: /LIMIT \?/) { Post.first }
+      # You can use +assert_sql+ assertion method for more advanced checks.
       #
-      def assert_queries(expected_count, matcher: nil, &block)
+      def assert_queries(expected_count, &block)
         ActiveRecord::Base.connection.materialize_transactions
 
-        queries = []
-        callback = lambda do |*, payload|
-          queries << payload[:sql] if %w[ SCHEMA TRANSACTION ].exclude?(payload[:name]) && (matcher.nil? || payload[:sql].match(matcher))
-        end
-        ActiveSupport::Notifications.subscribed(callback, "sql.active_record") do
+        counter = SQLCounter.new
+        ActiveSupport::Notifications.subscribed(counter, "sql.active_record") do
           result = _assert_nothing_raised_or_warn("assert_queries", &block)
+          queries = counter.log
           assert_equal expected_count, queries.size, "#{queries.size} instead of #{expected_count} queries were executed. Queries: #{queries.join("\n\n")}"
           result
         end
       end
 
       # Asserts that no SQL queries are executed in the given block.
+      #
+      #   assert_no_queries { post.comments }
+      #
       def assert_no_queries(&block)
         assert_queries(0, &block)
       end
+
+      # Asserts that the SQL queries executed in the given block match expected conditions.
+      #
+      #   # Check for exact number of queries
+      #   assert_sql(1) { Post.first }
+      #
+      #   # Check for any number of queries
+      #   assert_sql { Post.first }
+      #
+      # If the +:matcher+ option is provided, only queries that match the matcher are considered.
+      #
+      #   assert_sql(1, matcher: /LIMIT \?/) { Post.first }
+      #
+      # If the +:include_all+ option is provided, any queries (including transaction and schema related)
+      #   that match the matcher are considered.
+      #
+      #   assert_sql(include_all: true) { Post.columns }
+      #
+      def assert_sql(expected_count = nil, matcher: nil, include_all: false, &block)
```

Comment:
How do you feel about renaming the `:matcher` keyword argument to `:match`? It'd mirror the new `:match` keyword argument that [assert_raises](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Testing/Assertions.html#method-i-assert_raises) supports, and would correspond to the [Regexp#match](https://ruby-doc.org/core-3.0.1/Regexp.html#method-i-match) method name.

Reply:
👍